diff --git a/javalib/src/main/scala/java/lang/Byte.scala b/javalib/src/main/scala/java/lang/Byte.scala
index 0000000..0000000 100644
--- a/javalib/src/main/scala/java/lang/Byte.scala
+++ b/javalib/src/main/scala/java/lang/Byte.scala
@@ -39,13 +39,13 @@
   @inline override def doubleValue(): scala.Double =
     _value.toDouble

   @inline override def equals(that: Any): scala.Boolean =
-    that match {
+    (this eq that.asInstanceOf[AnyRef]) || (that match {
       case that: java.lang.Byte =>
         _value == that._value
       case _ =>
         false
-    }
+    })

   @inline override def hashCode(): Int =
     _value
diff --git a/javalib/src/main/scala/java/lang/Short.scala b/javalib/src/main/scala/java/lang/Short.scala
index 0000000..0000000 100644
--- a/javalib/src/main/scala/java/lang/Short.scala
+++ b/javalib/src/main/scala/java/lang/Short.scala
@@ -40,13 +40,13 @@
   @inline override def doubleValue(): scala.Double =
     _value.toDouble

   @inline override def equals(that: Any): scala.Boolean =
-    that match {
+    (this eq that.asInstanceOf[AnyRef]) || (that match {
       case that: java.lang.Short =>
         _value == that._value
       case _ =>
         false
-    }
+    })

   @inline override def hashCode(): scala.Int =
     _value
diff --git a/javalib/src/main/scala/java/lang/Character.scala b/javalib/src/main/scala/java/lang/Character.scala
index 0000000..0000000 100644
--- a/javalib/src/main/scala/java/lang/Character.scala
+++ b/javalib/src/main/scala/java/lang/Character.scala
@@ -23,13 +23,13 @@
   def charValue(): scala.Char =
     _value

   @inline override def equals(that: Any): scala.Boolean =
-    that match {
+    (this eq that.asInstanceOf[AnyRef]) || (that match {
       case that: Character =>
         _value == that._value
       case _ =>
         false
-    }
+    })

   @inline override def compareTo(that: Character): Int =
     Character.compare(_value, that._value)
diff --git a/javalib/src/main/scala/java/lang/Integer.scala b/javalib/src/main/scala/java/lang/Integer.scala
index 0000000..0000000 100644
--- a/javalib/src/main/scala/java/lang/Integer.scala
+++ b/javalib/src/main/scala/java/lang/Integer.scala
@@ -41,13 +41,13 @@
   @inline override def doubleValue(): scala.Double =
     _value.toDouble

   @inline override def equals(that: Any): scala.Boolean =
-    that match {
+    (this eq that.asInstanceOf[AnyRef]) || (that match {
       case that: Integer =>
         _value == that._value
       case _ =>
         false
-    }
+    })

   @inline override def hashCode(): scala.Int =
     _value
diff --git a/javalib/src/main/scala/java/lang/Float.scala b/javalib/src/main/scala/java/lang/Float.scala
index 0000000..0000000 100644
--- a/javalib/src/main/scala/java/lang/Float.scala
+++ b/javalib/src/main/scala/java/lang/Float.scala
@@ -42,18 +42,18 @@
   @inline override def doubleValue(): scala.Double =
     _value.toDouble

   @inline override def equals(that: Any): scala.Boolean =
-    that match {
+    (this eq that.asInstanceOf[AnyRef]) || (that match {
       case that: Float =>
         // As floats with different NaNs are considered equal,
         // use floatToIntBits instead of floatToRawIntBits
         val a = Float.floatToIntBits(_value)
         val b = Float.floatToIntBits(that._value)
         a == b

       case _ =>
         false
-    }
+    })

   @inline override def hashCode(): Int =
     Float.hashCode(_value)
