diff --git a/nir/src/main/scala/scala/scalanative/nir/Insts.scala b/nir/src/main/scala/scala/scalanative/nir/Insts.scala
index 586605f..322ef03 100644
--- a/nir/src/main/scala/scala/scalanative/nir/Insts.scala
+++ b/nir/src/main/scala/scala/scalanative/nir/Insts.scala
@@ -1,7 +1,13 @@
 package scala.scalanative
 package nir
 
+import scala.annotation.nowarn
+import scala.util.hashing.MurmurHash3
+
 sealed abstract class Inst extends Positioned {
+  self: Product =>
+  override lazy val hashCode = MurmurHash3.productHash(self): @nowarn
+
   final def show: String = nir.Show(this)
 }
 
@@ -22,7 +28,7 @@ object Inst {
       Let(fresh(), op, unwind)
   }
 
-  sealed abstract class Cf extends Inst
+  sealed abstract class Cf extends Inst with Product
   final case class Ret(value: Val)(implicit val pos: SourcePosition) extends Cf
   final case class Jump(next: Next)(implicit val pos: SourcePosition) extends Cf
   final case class If(value: Val, thenp: Next, elsep: Next)(implicit
diff --git a/nir/src/main/scala/scala/scalanative/nir/Types.scala b/nir/src/main/scala/scala/scalanative/nir/Types.scala
index 7d4d46c..a879f36 100644
--- a/nir/src/main/scala/scala/scalanative/nir/Types.scala
+++ b/nir/src/main/scala/scala/scalanative/nir/Types.scala
@@ -1,9 +1,14 @@
 package scala.scalanative
 package nir
 
+import scala.annotation.nowarn
+import scala.util.hashing.MurmurHash3
+
 import util.{unreachable, unsupported}
 
 sealed abstract class Type {
+  self: Product =>
+  override lazy val hashCode = MurmurHash3.productHash(self): @nowarn
 
   final def elemty(path: Seq[Val]): Type = (this, path) match {
     case (_, Seq()) =>
@@ -37,7 +42,7 @@ sealed abstract class Type {
 object Type {
 
   /** The type of an aggregate or primitive value. */
-  sealed abstract class ValueKind extends Type
+  sealed abstract class ValueKind extends Type with Product
 
   /** A primitive value type.
    *
@@ -117,7 +122,7 @@ object Type {
   final case class StructValue(tys: Seq[Type]) extends AggregateKind
 
   /** A reference type. */
-  sealed abstract class RefKind extends Type {
+  sealed abstract class RefKind extends Type with Product {
 
     /** The identifier of the class corresponding to this type. */
     final def className: Global.Top = this match {
@@ -173,7 +178,7 @@ object Type {
   ) extends RefKind
 
   /** Second-class types. */
-  sealed abstract class SpecialKind extends Type
+  sealed abstract class SpecialKind extends Type with Product
   case object Vararg extends SpecialKind
   case object Nothing extends SpecialKind
   case object Virtual extends SpecialKind
diff --git a/nir/src/main/scala/scala/scalanative/nir/Vals.scala b/nir/src/main/scala/scala/scalanative/nir/Vals.scala
index 98f7078..6970a0a 100644
--- a/nir/src/main/scala/scala/scalanative/nir/Vals.scala
+++ b/nir/src/main/scala/scala/scalanative/nir/Vals.scala
@@ -4,8 +4,13 @@ package nir
 import java.lang.Double.doubleToRawLongBits
 import java.lang.Float.floatToRawIntBits
 
+import scala.annotation.nowarn
+import scala.util.hashing.MurmurHash3
+
 /** A NIR value. */
 sealed abstract class Val {
+  self: Product =>
+  override lazy val hashCode = MurmurHash3.productHash(self): @nowarn
 
   /** The type of the value. */
   final def ty: Type = this match {
