diff --git a/compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala b/compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala
index c8d9a9e9..e8a9bd0f 100644
--- a/compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala
+++ b/compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala
@@ -167,8 +167,8 @@ private[semanticdb] object ExtractSemanticDB:
       schema = Schema.SEMANTICDB4,
       language = Language.SCALA,
       uri = Tools.mkURIstring(Paths.get(relPath(source, sourceRoot))),
-      text = if semanticdbText then String(source.content) else "",
-      md5 = internal.MD5.compute(String(source.content)),
+      text = if semanticdbText then String(source.content()) else "",
+      md5 = internal.MD5.compute(String(source.content())),
       symbols = symbolInfos,
       occurrences = occurrences,
       synthetics = synthetics,
diff --git a/compiler/src/dotty/tools/dotc/semanticdb/PPrint.scala b/compiler/src/dotty/tools/dotc/semanticdb/PPrint.scala
index f96a4839..b30ecfdb 100644
--- a/compiler/src/dotty/tools/dotc/semanticdb/PPrint.scala
+++ b/compiler/src/dotty/tools/dotc/semanticdb/PPrint.scala
@@ -328,14 +328,14 @@ private[semanticdb] class SyntheticPrinter(symtab: PrinterSymtab, source: Source
       /** get the line length of a given line */
       def lineLength(line: Int): Int =
         val isLastLine = source.lineToOffsetOpt(line).nonEmpty && source.lineToOffsetOpt(line + 1).isEmpty
-        if isLastLine then source.content.length - source.lineToOffset(line) - 1
+        if isLastLine then source.content().length - source.lineToOffset(line) - 1
         else source.lineToOffset(line + 1) - source.lineToOffset(line) - 1 // -1 for newline char
 
       val start = source.lineToOffset(range.startLine) +
         math.min(range.startCharacter, lineLength(range.startLine))
       val end = source.lineToOffset(range.endLine) +
         math.min(range.endCharacter, lineLength(range.endLine))
-      new String(source.content, start, end - start)
+      new String(source.content(), start, end - start)
 
 
   // def pprint(tree: s.Tree, range: Option[Range]): String =
diff --git a/presentation-compiler/src/main/dotty/tools/pc/CachingDriver.scala b/presentation-compiler/src/main/dotty/tools/pc/CachingDriver.scala
index 352157bd..84dc8ff6 100644
--- a/presentation-compiler/src/main/dotty/tools/pc/CachingDriver.scala
+++ b/presentation-compiler/src/main/dotty/tools/pc/CachingDriver.scala
@@ -52,7 +52,7 @@ class CachingDriver private (
       case _ => false
 
   override def run(uri: URI, source: SourceFile): List[Diagnostic] =
-    if !alreadyCompiled(uri, source.content) then previousDiags = super.run(uri, source)
+    if !alreadyCompiled(uri, source.content()) then previousDiags = super.run(uri, source)
     lastCompiledURI = uri
     previousDiags
 
diff --git a/presentation-compiler/src/main/dotty/tools/pc/completions/OverrideCompletions.scala b/presentation-compiler/src/main/dotty/tools/pc/completions/OverrideCompletions.scala
index 3d9f48fd..ef3e83a7 100644
--- a/presentation-compiler/src/main/dotty/tools/pc/completions/OverrideCompletions.scala
+++ b/presentation-compiler/src/main/dotty/tools/pc/completions/OverrideCompletions.scala
@@ -120,7 +120,7 @@ object OverrideCompletions:
           shouldMoveCursor = true,
           config,
           autoImportsGen,
-          indexedContext.ctx.compilationUnit.source.content
+          indexedContext.ctx.compilationUnit.source.content()
             .startsWith("o", start)
         )
       )
diff --git a/presentation-compiler/src/main/dotty/tools/pc/utils/InteractiveEnrichments.scala b/presentation-compiler/src/main/dotty/tools/pc/utils/InteractiveEnrichments.scala
index 1f2e10f2..25ebcd15 100644
--- a/presentation-compiler/src/main/dotty/tools/pc/utils/InteractiveEnrichments.scala
+++ b/presentation-compiler/src/main/dotty/tools/pc/utils/InteractiveEnrichments.scala
@@ -193,7 +193,7 @@ object InteractiveEnrichments extends CommonMtagsEnrichments:
       sym.sourcePos.span.exists && {
         val source = ctx.source
         if (source ne sym.source) && source.path == sym.source.path then
-          !source.content.startsWith(
+          !source.content().startsWith(
             sym.decodedName.toString(),
             sym.sourcePos.span.point
           )
